Solutions/Auth0/Data Connectors/Auth0Connector/main.py

```diff
@@ -76,45 +68,98 @@ def __init__(self, domain, api_path, client_id, client_secret, audience):
         self.uri = self.domain + self.api_path
         self.token = None
         self.header = None
-
-    def get_log_events(self, config: dict) -> Tuple[str, List]:
+        self.retry = retry  
+
+    """This method is used to process and post the results to Log Analytics Workspace
+        Returns:
+            last_log_id : last processed eventId
+            events: last processed Events
+    """
+    def get_log_events(self, script_start_time, config: dict) -> Tuple[str, List]:
         self.token = self._get_token()
         logging.info(f'Token provided.')
         self.header = self._get_header()
         last_log_id = self._get_last_log_id(config)
         #last_log_id = "90020230126121002244690048186607762971591195832157732866"
         logging.info(f'\tLast log id extracted: {last_log_id}.')
         if last_log_id is None:
-            return '', []
+            #return '', []
+            self.update_statemarker_file(config, '', [])
+            return
         # first request
         params = {'from': last_log_id, 'take': '100'}
-        resp = requests.get(self.uri, headers=self.header, params=params)
+        count = 0
+        error=True
+        while error:
+            try:
+                error=False
+                resp = requests.get(self.uri, headers=self.header, params=params)
+            except Exception as err:
+                error = True
+                count+=1
+                logging.error("Something wrong. Exception error text: {}".format(err))
+                if count > self.retry:
+                    logging.error("Exceeded maximum Retries")
+                    break
         logging.info('\tFirst request executed.')
         if not resp.json():
-            return last_log_id, []
+            #return last_log_id, []
+            self.update_statemarker_file(config, last_log_id, [])
+            return
         events = resp.json()
         logging.info('\t response object : {events}')
+        events.sort(key=lambda item: item['date'], reverse=True)
+        last_log_id = events[0]['log_id']
+        for el in events:
+            self.sentinel.send(el)
+        self.sentinel.flush()
+        logging.info('Events sent to Sentinel.')
+        self.update_statemarker_file(config, last_log_id , events)
+
         if "Link" in resp.headers :
             next_link = resp.headers['Link']
             next_uri = next_link[next_link.index('<') + 1:next_link.index('>')]
             page_num = 1
-            while resp.json():
-                resp = requests.get(next_uri, headers=self.header)
+            while resp.json() and len(events)!=0:
+                count = 0
+                error=True
+                while error:
+                    try:
+                        error=False
+                        resp = requests.get(next_uri, headers=self.header)
```

Comment:
Could  you please check status code responses  and handle those in code for retry i.e. 429, 500, 502, 503, 504 and other status codes i.e. 401 ,400 etc.

Reply:
I have added backlog item to incorporate these changes.